src/resolve/resolved-dns-cache.c

```diff
@@ -18,6 +18,9 @@
 /* We never keep any item longer than 2h in our cache */
 #define CACHE_TTL_MAX_USEC (2 * USEC_PER_HOUR)
```

Comment:
(I commented in a resolved thread but i'm not sure how to un-resolve it, so adding a new review)

This (clamping dns TTL values to 2 hours, but also keeping "stale" records after we think they "expire") still makes no sense to me. Not sure if @poettering agrees with me or not.

The only benefit I can see for clamping TTL is to reduce how long records stay in our cache. However, if we are keeping the record in our cache past the clamped TTL, there really wasn't any point in clamping it at all, in fact clamping the TTL actually makes everything worse.

Reply:
so my thinking was that overly long dns cache times don't really make sense on a client. we'd rather ask too often then too seldom. we are not a load-bearing DNS server for a bunch of clients, we are basically just an improvemet over the cache-less in-glibc resolver. Hence in the interest in being able to somewhat gracefully handle DNS misconfiguration i figured it makes sense to revalidate all entries after 2h no matter what, even if they are supposedly still valid for a year.

dunno. i tink this should be discussed separately from this PR.

Reply:
so i looked into this in more detail. I think the confusion comes from the fact how CACHE_TTL_MAX_USEC is applied. 

If the stale logic is enabled it means that if a request comes in later than 2h after caching we'll try to refresh every record. If it works, then great, we'll return the refreshed RRset. But if it doesn't, then we'll return the stale data, up to the configured stale rentation timeout. 

If the stale logic is disabled it means that if a request comes in later than 2h after caching we'll try to refresh it. If it works, great. If not, we'll return an error.

I think this is a reasonable approach I think. The question of course is what is more important to people: honour superlong ttls because it allows taping over connectivity problems (A), or limit superlong ttls because it allows taping over DNS configuration problems (B). I think when figuring out the answer to this one should take into consideration whether one is looking at a DNS resolver such as resolved that is on the end-user's system such as systemd-resolved or whether one is looking at a DNS resolver at an ISP or so. An ISP DNS resolver will have many upstreams a client DNS resolver pretty much only one it strictly relies on.